lib/galaxy/workflow/scheduling_manager.py

```diff
@@ -326,9 +338,55 @@ def __schedule(self, workflow_scheduler_id, workflow_scheduler):
             if not self.monitor_running:
                 return
 
+    def __attempt_materialize(self, workflow_invocation, session) -> bool:
+        try:
+            inputs_to_materialize = workflow_invocation.inputs_requiring_materialization()
+            for input_to_materialize in inputs_to_materialize:
+                hda = input_to_materialize.hda
+                user = RequestUser(user_id=workflow_invocation.history.user_id)
+                task_request = MaterializeDatasetInstanceTaskRequest(
+                    user=user,
+                    history_id=workflow_invocation.history.id,
+                    source="hda",
+                    content=hda.id,
+                    validate_hashes=True,
+                )
+                materialized_okay = self.app.hda_manager.materialize(task_request, in_place=True)
+                if not materialized_okay:
+                    workflow_invocation.fail()
+                    workflow_invocation.add_message(
+                        InvocationFailureDatasetFailed(
+                            workflow_step_id=input_to_materialize.input_dataset.workflow_step.id,
+                            reason=FailureReason.dataset_failed,
+                            hda_id=hda.id,
+                        )
+                    )
+                    session.add(workflow_invocation)
+                    session.commit()
+                    return False
+
+            # place back into ready and let it proceed normally on next iteration?
+            workflow_invocation.set_state(model.WorkflowInvocation.states.READY)
+            session.add(workflow_invocation)
+            session.commit()
+            return True
+        except Exception as e:
+            log.exception(f"Failed to materialize dataset for workflow {workflow_invocation.id} - {e}")
+            workflow_invocation.fail()
+            failure = InvocationUnexpectedFailure(reason=FailureReason.unexpected_failure, details=str(e))
+            workflow_invocation.add_message(failure)
+            session.add(workflow_invocation)
+            session.commit()
+        return False
+
     def __attempt_schedule(self, invocation_id, workflow_scheduler):
         with self.app.model.context() as session:
             workflow_invocation = session.get(model.WorkflowInvocation, invocation_id)
+            if workflow_invocation.state == workflow_invocation.states.REQUIRES_MATERIALIZATION:
+                if not self.__attempt_materialize(workflow_invocation, session):
+                    return None
+                if self.app.config.workflow_scheduling_separate_materialization_iteration:
+                    return None
```

Comment:
Does this have any measurable effect ? I would imagine materialization of even a small dataset takes so much longer than scheduling a workflow. As I understand this, we now spend a lot of time materializing a number of inputs (unless deferred), and then based on this setting we proceed or not the schedule the workflow.